python/pyspark/sql/types.py

```diff
@@ -507,20 +503,52 @@ def add(self, field, data_type=None, nullable=True, metadata=None):
             else:
                 data_type_f = data_type
             self.fields.append(StructField(field, data_type_f, nullable, metadata))
-            self.names.append(field)
-        self._needSerializeAnyField = any(f.needConversion() for f in self.fields)
         return self
 
+    @property
+    def names(self):
+        """Return a list of the field names"""
+        return [field.name for field in self]
+
+    @property
+    def _needSerializeAnyField(self):
+        """Determine if any field needs conversion"""
+        return any(field.needConversion() for field in self)
+
+    def __iter__(self):
+        """Iterate the fields"""
+        return iter(self.fields)
+
+    def __len__(self):
+        """Return the number of fields."""
+        return len(self.fields)
+
+    def __getitem__(self, key):
+        """Access fields by name or slice."""
+        if isinstance(key, str):
+            _dict_fields = {field.name: field for field in self}
```

Comment:
The dictionary will only apply if they access via a `str`, so it can only return a single field.

It did just occur to me that using an actual slice object will return a `list` of `StructField` objects instead of a `StructType` object in this code currently.  I'll go fix that now.